### PR TITLE
kv: add partial implementation of new memdb

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve?
This is the first patchset of https://github.com/pingcap/tidb/pull/18650. Please see the full description in the original PR page.

Contains the Red-Black Tree implementation. The missing parts will be added in other patches

### Release note 

-  No release note

### CODE REVIEW

File: kv/new_memdb.go [hunk 1]
```diff
@@ -0,0 +1,766 @@
...
+	// NOTE: Since the root node is guaranteed black, then we
+	// are also going to stop if we are the child of the root
+
+	for x.addr != db.root {
+		xup := x.getUp(db)
+		if xup.isBlack() {
+			break
+		}
+
+		xgrandup := db.getNode(xup.up)
```
Comment (coocood) [inline kv/new_memdb.go:354]:
```suggestion
		xgrandup := xup.getUp(db)
```

File: kv/new_memdb.go [hunk 2]
```diff
@@ -0,0 +1,766 @@
...
+		// if our parent is on the left side of our grandparent
+		if x.up == xgrandup.left {
+			// get the right side of our grandparent (uncle?)
+			y = db.getNode(xgrandup.right)
```
Comment (coocood) [inline kv/new_memdb.go:358]:
```suggestion
			y = xgrandup.getRight(db)
```

File: kv/new_memdb.go [hunk 3]
```diff
@@ -0,0 +1,766 @@
...
+			if y.isRed() {
+				// make our parent black
+				xup.setBlack()
+				// make our uncle black
+				y.setBlack()
+				// make our grandparent red
+				xgrandup.setRed()
+				// now consider our grandparent
+				x = db.getNode(xup.up)
```
Comment (coocood) [inline kv/new_memdb.go:367]:
```suggestion
				x = xup.getUp(db)
```

File: kv/new_memdb.go [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+			} else {
+				// if we are on the right side of our parent
+				if x.addr == xup.right {
+					// Move up to our parent
+					x = x.getUp(db)
+					db.leftRotate(x)
+					xup = x.getUp(db)
+					xgrandup = db.getNode(xup.up)
```
Comment (coocood) [inline kv/new_memdb.go:375]:
```suggestion
					xgrandup = xup.getUp(db)
```

**Comment by coocood** [comment]
LGTM

File: kv/new_memdb_arena.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	"encoding/binary"
+	"math"
+)
+
+var (
+	nullAddr = memdbArenaAddr{math.MaxUint32, math.MaxUint32}
+	endian   = binary.LittleEndian
+)
+
+type memdbArenaAddr struct {
```
Comment (tiancaiamao) [inline kv/new_memdb_arena.go:26]: This file is here just to make the build pass?

**Pull request merged**